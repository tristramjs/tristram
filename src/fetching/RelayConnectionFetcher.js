/* @flow */
import type { RawSiteMapData } from '../types/sitemap';

import type { Fetcher, FetcherProps } from './index';

require('es6-promise').polyfill();
require('isomorphic-fetch');

type Props<T> = FetcherProps<T> & {
	chunkSize?: number,
	getConnection: GetConnection<T>,
	query: string,
	logErrors?: boolean,
	maxRetries?: number,
};

type GetConnection<T> = <T>(x: Object) => GraphQlConnection<T>;
type GraphQlConnection<T> = { edges: { node: T, cursor: string }[], pageInfo: { hasNextPage: boolean } };

export default class RelayConnection<T> implements Fetcher {
	url: string;
	query: string;
	chunkSize: number;
	transformResult: T => RawSiteMapData;
	logErrors: boolean;
	maxRetries: number;
	getConnection: GetConnection<T>;
	variables: {
		first: number,
		after?: string,
	} = {
		first: 10,
	};

	constructor({
		url, getConnection, transformResult, query, chunkSize, logErrors, maxRetries,
	}: Props<T>) {
		this.url = url;
		this.getConnection = getConnection;
		this.transformResult = transformResult;
		this.query = query;
		this.maxRetries = maxRetries || 3;
		this.chunkSize = chunkSize || 100;
		this.logErrors = logErrors || false;
	}

	async* getData(): AsyncGenerator<RawSiteMapData[], void, void> {
		let hasNextPage;

		do {
			const data: GraphQlConnection<T> = await this.fetchConnection();

			yield data.edges.map(edge => this.transformResult(edge.node));
			({ hasNextPage } = data.pageInfo);

			this.variables = {
				first: this.chunkSize,
				after: data.edges[data.edges.length - 1].cursor,
			};
		} while (hasNextPage);
	}

	async fetchConnection(): Promise<GraphQlConnection<T>> {
		const {
			url, query, variables, getConnection, logErrors,
		} = this;
		let retries = 0;
		const options = {
			method: 'POST',
			headers: {
				'content-type': 'application/json',
			},
			body: JSON.stringify({ query, variables }),
		};

		while (retries < this.maxRetries) {
			const res = await fetch(url, options);
			const data = await res.json();

			if (data.errors && Array.isArray(data.errors)) {
				if (logErrors) {
					// eslint-disable-next-line no-console
					console.log(
						`RelayConnectionFetcher: Received errors in connection:\n"${data.errors
							.map(err => (err.message ? err.message : '<no error message provided>'))
							.join('\n')}"`
					);
				}

				retries = retries + 1;
			} else {
				return getConnection(data);
			}
		}

		throw new Error(`Error fetching connection, tried ${retries} times.`);
	}
}
