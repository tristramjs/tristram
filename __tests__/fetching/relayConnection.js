/* @flow */
import RelayConnectionFetcher from '../../src/fetching/RelayConnectionFetcher';

describe('fetching.gqlConnection', () => {
	it('should fetch the whole connection in chunks', async () => {
		const fetcher = new RelayConnectionFetcher({
			url: 'http://test.com/graphql',
			query: /* GraphQL */ `query Test($first: Int!, $after: String) {
				viewer {
					allTestItems(first: $first, after: $after) {
						pageInfo {
							hasNext
						}
						edges {
							cursor
							node {
								id
							}
						}
					}
				}
			}`,
			getConnection: conn => conn.data.viewer.allTestItems,
			transformResult: ({ id }) => ({ loc: id }),
		});

		for await (const item of fetcher.getData()) {
			expect(item).toMatchSnapshot();
		}
	});
});
