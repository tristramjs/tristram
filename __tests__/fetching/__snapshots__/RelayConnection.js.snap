// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetching.gqlConnection should fetch the whole connection in chunks 1`] = `
Array [
  Object {
    "loc": "bla",
  },
]
`;

exports[`fetching.gqlConnection should fetch the whole connection in chunks 2`] = `
Array [
  Object {
    "loc": "blerb",
  },
]
`;
